### PR TITLE
Playlist feature: add 'Shuffle playlist' sidebar action

### LINKED ISSUE #6988: Shuffle function for all playlists
Reported by: **[daschuer](https://github.com/daschuer)**
Date: 2013-04-16T06:30:36Z
Status: New
Importance: Wishlist
Launchpad Issue: [lp1169435](https://bugs.launchpad.net/bugs/1169435)

---

Hi all,

This is simply a feature suggestion...no need for reply:

The Auto DJ system has a "shuffle" feature, but playlists and crates do not.  I'm setting up for an event and was sent a list of around 90 songs to have handy.  The list came alphabetized by artist.  So, I added them all to a playlist.  But, I also have 3 songs that need to be cued at specific times.

What I notice is that the playlist ended up in the order that I inserted them (hence in order of artist name) and I had no option to pre-shuffle them.  Therefore, when I do my event, I'll have to add the playlist to the Auto DJ, shuffle, and then add my cues in the their proper place (because shuffle would move them if they were already in the playlist.)

Not any huge deal...I can work around it, but I'm super busy during these things (as I'm sure you all can relate to) and it would save me time.  Just an idea.  Maybe there's another work-around?

Thanks!
Kendell

### PR DESCRIPTION
**Pull request by ronso0**
Closes #6988 

* shuffle the entire playlist
* if the playlist is selected / content visible in tracks view and at least 2 tracks are selected: shuffle only selected tracks

### CODE REVIEW

**Review by daschuer — COMMENTED** [review]
LGTM, I have just left a suggestion. Please give a hint for merge.

File: src/library/trackset/playlistfeature.cpp [hunk 1]
```diff
@@ -222,6 +234,49 @@ QString PlaylistFeature::fetchPlaylistLabel(int playlistId) {
...
+                    m_pLibraryWidget->getActiveView());
+            if (view != nullptr) {
+                selection = view->selectionModel()->selectedIndexes();
+            }
+        }
+        m_pPlaylistTableModel->shuffleTracks(selection, QModelIndex());
+    } else {
+        // Create a temp model so we don't need to select the playlist
+        // in the persistent model in order to shuffle it
+        QScopedPointer<PlaylistTableModel> pPlaylistTableModel(
```
[the remark this answers is not in the document]
Comment (ronso0) [reply src/library/trackset/playlistfeature.cpp]: Tbh I didn't think about it in this case. Ime my PRs sit for too long when I optimize too much en passant.
However, there are 9 occurences of this pattern in the playlist features and I can of course change them to use unique_ptr

**Comment by ronso0** [comment]
Ready for merge.

File: src/library/trackset/playlistfeature.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (daschuer) [reply src/library/trackset/playlistfeature.cpp]: Ups, I did not notice how that there so many occurrences. Thank you very much for fixing them all.

**Review by daschuer — APPROVED** [review]
LGTM, thank you.

**Pull request merged**